### ISSUE TITLE
2021-12-7

### ISSUE THREAD

**Issue by ssfc** [ssfc/wangdao#677]
1011.Knowledge distillation: model compression; (2021-11-22)
1022.Adaptive Computation: Skipping some layers in neural networks; Skipping some channels in neural networks; Skipping less important pixels of the current input images; (2021-11-22)
1033.Deep Supervision: additional supervision is added to train the hidden layers directly; (2021-11-22)
104III.Self Distillation
1052.Figure 2; (i) A ResNet has been divided into four sections according to their depth; (what is usage of bottleneck??) (ii) Additional bottleneck and fully connected layers are set after each section, which constitutes multiple classifiers; (iii) All of the classifiers can be utilized independently, with different accuracy and response time; (iv) Each classifier is trained under three kinds of supervision as depicted; (v) Parts under the dash line can be removed in inference; (2021-11-23)
1063.construct the self distillation framework; (1) target convolutional neural network is divided into several shallow sections according to its depth and original structure; (2) a classifier, combined with a bottleneck [14] layer and a fully connected layer which are only utilized in training and can be removed in inference, is set after each shallow section; REASON, adding the bottleneck layer is to mitigate the impacts between each shallow classifier, and to add L2 loss from hints; (3) all the shallow sections with corresponding classifiers are trained as student models via distillation from the deepest section, which can be conceptually regarded as the teacher model; (2021-11-23)
1074.Loss Source 1: Cross entropy loss from labels to not only the deepest classifier, but also all the shallow classifiers; (2021-11-23)
1085.Loss Source 2: KL (Kullback-Leibler) divergence loss under teacher’s guidance; KL divergence is computed using softmax outputs between students and teachers; (2021-11-23)
1096.Loss Source 3: L2 loss from hints; computation of the L2 loss between features maps of the deepest classifier and each shallow classifier; (2021-11-23)
